core/services/synchronization/telemetry_ingress_client.go

```diff
@@ -0,0 +1,156 @@
+
+type telemetryIngressClient struct {
+	utils.StartStopOnce
+	url             *url.URL
+	ks              keystore.CSAKeystoreInterface
+	serverPubKeyHex string
+	wsrpcClient     pb.TelemClient
+	logging         bool
+
+	mu      *sync.RWMutex
+	isReady bool
+	wgDone  sync.WaitGroup
+	chDone  chan struct{}
+}
+
+// NewTelemetryIngressClient returns a client backed by wsrpc that
+// can send telemetry to the telemetry ingress server
+func NewTelemetryIngressClient(url *url.URL, serverPubKeyHex string, ks keystore.CSAKeystoreInterface, logging bool) TelemetryIngressClient {
+	return &telemetryIngressClient{
+		url:             url,
+		ks:              ks,
+		serverPubKeyHex: serverPubKeyHex,
+		logging:         logging,
+		mu:              new(sync.RWMutex),
+	}
+}
+
+// Start connects the wsrpc client to the telemetry ingress server
+func (tc *telemetryIngressClient) Start() error {
+	return tc.StartOnce("TelemetryIngressClient", func() error {
+		privkey, err := tc.getCSAPrivateKey()
+		if err != nil {
+			return err
+		}
+
+		tc.connect(privkey)
+
+		return nil
+	})
+}
+
+// Close disconnects the wsrpc client from the ingress server
+func (tc *telemetryIngressClient) Close() error {
+	return tc.StopOnce("TelemetryIngressClient", func() error {
+		close(tc.chDone)
+		tc.wgDone.Wait()
+		return nil
+	})
+}
+
+func (tc *telemetryIngressClient) connect(clientPrivKey []byte) {
+	tc.wgDone.Add(1)
+
+	go func() {
+		defer tc.wgDone.Done()
+
+		serverPubKey := keys.FromHex(tc.serverPubKeyHex)
+
+		conn, err := wsrpc.Dial(tc.url.String(), wsrpc.WithTransportCreds(clientPrivKey, serverPubKey))
+		if err != nil {
+			logger.Errorf("Error connecting to telemetry ingress server: %v", err)
+			return
+		}
+		defer conn.Close()
+
+		// Initialize a new wsrpc client caller
+		// This is used to call RPC methods on the server
+		tc.mu.Lock()
+		tc.wsrpcClient = pb.NewTelemClient(conn)
+		tc.isReady = true
+		tc.mu.Unlock()
+
+		// Wait for close
+		<-tc.chDone
+
+	}()
+}
+
+// getCSAPrivateKey gets the client's CSA private key
+func (tc *telemetryIngressClient) getCSAPrivateKey() (privkey []byte, err error) {
+	// Fetch the client's public key
+	keys, err := tc.ks.ListCSAKeys()
+	if err != nil {
+		return privkey, err
+	}
+	if len(keys) < 1 {
+		return privkey, errors.New("CSA key does not exist")
+	}
+
+	privkey, err = tc.ks.Unsafe_GetUnlockedPrivateKey(keys[0].PublicKey)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	return privkey, nil
+}
+
+// Send sends telemetry to the ingress server using wsrpc if the client is ready
+func (tc *telemetryIngressClient) Send(ctx context.Context, telemetry []byte, contractAddr common.Address) {
+	telemReq := &pb.TelemRequest{Telemetry: telemetry, Address: contractAddr.String()}
+
+	tc.mu.RLock()
+	defer tc.mu.RUnlock()
+
+	if !tc.isReady {
+		logger.Error("Could not send telemetry, client is not ready")
+		return
+	}
+
+	go func() {
```

Comment:
I would implement this by sending a channel to your worker go routine above, rather than sharing the `tc.wsrpcClient`. We should have plenty of examples of how to do this in the codebase, ping me if you need some guidance.

Reply:
Added a commit that switches the telem ingress client to instead use a buffered channel (pretty much used the implementation found in `explorer.go`)

Didn't rebase the commit for better change visibility, but let me know if I should